sql/sql_class.cc

```diff
@@ -6465,6 +6469,36 @@ int THD::decide_logging_format(TABLE_LIST *tables)
           lex->set_stmt_accessed_table(trans ? LEX::STMT_WRITES_TRANS_TABLE :
                                                LEX::STMT_WRITES_NON_TRANS_TABLE);
 
+#ifndef EMBEDDED_LIBRARY
+        if (rgi_slave && rgi_slave->is_parallel_exec &&
+            likely(rgi_slave->gtid_ev_flags2 &
+                   Gtid_log_event::FL_TRANSACTIONAL) &&
+            unlikely(!trans) &&
+            rgi_slave->parallel_entry->unsafe_rollback_marker_sub_id.load(
+                std::memory_order_relaxed) < rgi_slave->gtid_sub_id)
+        {
+          DBUG_ASSERT(!transaction->all.modified_non_trans_table);
+
+          // slave altered a transactional engine to non-transactional
+          struct rpl_parallel_entry *e= rgi_slave->parallel_entry;
+
+          mysql_mutex_lock(&e->LOCK_parallel_entry);
+          if (!e->force_abort)
+          {
+            e->unsafe_rollback_marker_sub_id=
+              std::max(rgi_slave->gtid_sub_id, e->unsafe_rollback_marker_sub_id.
+                       load(std::memory_order_relaxed));
+          }
+          else
+          {
+            // todo: issue a warning/errror
+            mysql_mutex_unlock(&e->LOCK_parallel_entry);
+            DBUG_RETURN(-1);
+          }
+          mysql_mutex_unlock(&e->LOCK_parallel_entry);
+        }
+#endif
```

Comment:
1. This code surely does not belong here? decide_logging_format() has the task to select between logging in statement or row. This logic seems to be about tracking non-transactional changes for faster stop of parallel replication, completely unrelated. And what if binlogging isn't enabled on the slave, is this code then not called?
2. Parallel replication does not support different engine on master and slave (I think the purpose of this code is to handle a table that is eg. InnoDB on the master but altered to MyISAM on the slave). This situation will cause failures in many different ways, not just for fast STOP SLAVE. And even this case seems not correctly handled by this code; an earlier event group could check unsafe_rollback_marker_sub_id and decide to roll back just before this code runs and sets the new value of unsafe_rollback_marker_sub_id ? It is a valid task to make parallel replication safe for replicating between different engines on master and slave, but it does not make sense to implement a partial solution.

Reply:
Completely agree with p.2. (I was against doing that for the very same reason; however we pushed it to satisfy @montywi 's review notes on the corporate slack).

To p.1, indeed the changes  in this function are ineffectual in `log_bin = ON` mode. That's a good point of yours.